File: frontend/src/types/work.ts
// Types for normalized work data.

export type Priority = "low" | "normal" | "high" | "urgent";

export interface User {
  id: string;
  email: string;
  full_name: string;
  display_name: string;
  timezone: string;
  is_active: boolean;
  created_at: string;
}

export interface Organization {
  id: string;
  name: string;
  slug: string;
}

export interface MeResponse {
  user: User;
  organization: Organization;
  role: string;
}

export interface TokenPair {
  access_token: string;
  refresh_token: string;
  token_type: string;
  expires_in: number;
}

export interface EmailItem {
  id: string;
  source: string;
  subject: string;
  body_preview: string;
  sender_email: string;
  sender_name: string;
  to_recipients: string[];
  received_at: string | null;
  is_read: boolean;
  has_attachments: boolean;
  priority: Priority;
  needs_reply: boolean;
  is_important: boolean;
  external_id?: string;
  project_id?: string | null;
  web_link?: string | null;
}

export interface CalendarEvent {
  id: string;
  source: string;
  subject: string;
  body_preview: string;
  starts_at: string;
  ends_at: string;
  is_all_day: boolean;
  location: string;
  organizer_email: string;
  attendees: { email?: string; name?: string }[];
  is_online_meeting: boolean;
  join_url?: string | null;
  response_status: string;
}

export interface MessageItem {
  id: string;
  source: string;
  conversation_id: string;
  conversation_name: string;
  sender_email: string;
  sender_name: string;
  body: string;
  direction: string;
  sent_at: string | null;
  is_mention: boolean;
}

export interface TaskItem {
  id: string;
  source: string;
  external_id: string;
  title: string;
  description: string;
  status: string;
  priority: Priority;
  due_at: string | null;
  assignee_name?: string | null;
  project_id?: string | null;
  progress: number;
  is_overdue: boolean;
  web_link?: string | null;
}

export interface Commitment {
  id: string;
  direction: "i_owe" | "owed_to_me";
  status: string;
  text: string;
  counterparty_email?: string | null;
  counterparty_name?: string | null;
  due_at: string | null;
  confidence: number;
  source: string;
  acknowledged: boolean;
}

export interface TimelineItem {
  id: string;
  source: string;
  type: string;
  timestamp: string | null;
  title: string;
  summary: string;
  people: string[];
  project?: string | null;
  priority: Priority;
  metadata: Record<string, unknown>;
}

export interface AttentionCounts {
  requiring_action: number;
  overdue_tasks: number;
  important_emails: number;
  upcoming_meetings: number;
}

export interface Dashboard {
  greeting: string;
  generated_at: string;
  counts: AttentionCounts;
  urgent: TimelineItem[];
  upcoming_events: CalendarEvent[];
  waiting_for: Commitment[];
  you_owe: Commitment[];
  important_emails: EmailItem[];
  overdue_tasks: TaskItem[];
  today_tasks: TaskItem[];
  integrations_warning: string[];
  is_development_data: boolean;
}

export interface IntegrationOverview {
  provider: string;
  display_name: string;
  description: string;
  connected: boolean;
  status: string;
  configured: boolean;
  account_label?: string | null;
  last_synced_at?: string | null;
  last_error?: string | null;
  is_mock: boolean;
  connect_url?: string | null;
}

export interface SyncState {
  provider: string;
  resource: string;
  status: string;
  last_synced_at: string | null;
  last_error: string | null;
  items_synced: number;
  consecutive_failures: number;
}

export interface Paginated<T> {
  items: T[];
  total: number;
  page: number;
  page_size: number;
}

File: backend/app/schemas/work.py
"""Schemas for normalized work data: emails, events, messages, tasks, timeline."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class EmailOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source: str
    external_id: str
    subject: str
    body_preview: str
    sender_email: str
    sender_name: str
    to_recipients: list[str] = Field(default_factory=list)
    received_at: datetime | None = None
    is_read: bool
    has_attachments: bool
    priority: str
    needs_reply: bool
    is_important: bool
    project_id: uuid.UUID | None = None
    web_link: str | None = None


class CalendarEventOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source: str
    external_id: str
    subject: str
    body_preview: str
    starts_at: datetime
    ends_at: datetime
    is_all_day: bool
    location: str
    organizer_email: str
    attendees: list[dict[str, Any]] = Field(default_factory=list)
    is_online_meeting: bool
    join_url: str | None = None
    response_status: str


class MessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source: str
    external_id: str
    conversation_id: str
    conversation_name: str
    sender_email: str
    sender_name: str
    body: str
    direction: str
    sent_at: datetime | None = None
    is_mention: bool


class TaskOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source: str
    external_id: str
    title: str
    description: str
    status: str
    priority: str
    due_at: datetime | None = None
    completed_at: datetime | None = None
    assignee_email: str | None = None
    assignee_name: str | None = None
    project_id: uuid.UUID | None = None
    progress: int
    is_overdue: bool
    web_link: str | None = None


class CommitmentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    direction: str
    status: str
    text: str
    counterparty_email: str | None = None
    counterparty_name: str | None = None
    due_at: datetime | None = None
    confidence: float
    source: str
    source_ref: str | None = None
    project_id: uuid.UUID | None = None
    acknowledged: bool


class TimelineItem(BaseModel):
    """A unified, provider-agnostic timeline entry."""

    id: str
    source: str
    type: str
    timestamp: datetime | None
    title: str
    summary: str = ""
    people: list[str] = Field(default_factory=list)
    project: str | None = None
    priority: str = "normal"
    ref_id: uuid.UUID | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class TimelineResponse(BaseModel):
    items: list[TimelineItem]
    next_cursor: str | None = None


class Paginated(BaseModel):
    items: list[Any]
    total: int
    page: int
    page_size: int